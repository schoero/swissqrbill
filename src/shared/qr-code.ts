import { cleanData } from "swissqrbill:shared:cleaner.js";
import { qrcodegen } from "swissqrbill:shared:qr-code-generator.js";
import { validateData } from "swissqrbill:shared:validator.js";
import { getReferenceType, mm2pt } from "swissqrbill:utils";

import type { Data } from "swissqrbill:shared:types.js";


export function generateQRData(data: Data): string {

  const cleanedData = cleanData(data);

  validateData(cleanedData);

  const amount = cleanedData.amount?.toFixed(2);
  const reference = getReferenceType(cleanedData.reference);

  const qrData: string[] = [
    "SPC",                                                               // Swiss Payments Code
    "0200",                                                              // Version
    "1",                                                                 // Coding Type UTF-8
    cleanedData.creditor.account ?? "",                                  // IBAN
    "S",                                                                 // Address Type
    cleanedData.creditor.name,                                           // Name
    cleanedData.creditor.address,                                        // Address
    cleanedData.creditor.buildingNumber                                  // Building number
      ? `${cleanedData.creditor.buildingNumber}`
      : "",
    `${cleanedData.creditor.zip}`,                                       // Zip
    cleanedData.creditor.city,                                           // City
    cleanedData.creditor.country,                                        // Country
    "",                                                                  // 1x Empty
    "",                                                                  // 2x Empty
    "",                                                                  // 3x Empty
    "",                                                                  // 4x Empty
    "",                                                                  // 5x Empty
    "",                                                                  // 6x Empty
    "",                                                                  // 7x Empty
    amount ?? "",                                                        // Amount
    cleanedData.currency,                                                // Currency
    ...cleanedData.debtor
      ? [
        "S",                                                             // Address Type
        cleanedData.debtor.name,                                         // Name
        cleanedData.debtor.address,                                      // Address
        cleanedData.debtor.buildingNumber                                // Building number
          ? `${cleanedData.debtor.buildingNumber}`
          : "",
        `${cleanedData.debtor.zip}`,                                     // Zip
        cleanedData.debtor.city,                                         // City
        cleanedData.debtor.country ?? ""                                 // Country
      ]
      : [
        "",                                                              // Empty address type
        "",                                                              // Empty name
        "",                                                              // Empty address
        "",                                                              // Empty building number
        "",                                                              // Empty zip field
        "",                                                              // Empty city field
        ""                                                               // Empty country
      ],
    reference,                                                           // Reference type
    cleanedData.reference ?? "",                                         // Reference
    cleanedData.message ?? "",                                           // Unstructured message
    "EPD",                                                               // End of payment data
    cleanedData.additionalInformation ?? "",                             // Additional information
    ...cleanedData.av1                                                   // Alternative scheme 1
      ? [
        cleanedData.av1
      ]
      : [],
    ...cleanedData.av2                                                   // Alternative scheme 2
      ? [
        cleanedData.av2
      ]
      : []
  ];

  return qrData.join("\n");

}


export function renderQRCode(data: Data, size: number, renderBlockFunction: (x: number, y: number, blockSize: number) => void): void {

  const qrData = generateQRData(data);
  const eci = qrcodegen.QrSegment.makeEci(26);
  const segments = qrcodegen.QrSegment.makeSegments(qrData);
  const qrCode = qrcodegen.QrCode.encodeSegments([eci, ...segments], qrcodegen.QrCode.Ecc.MEDIUM, 10, 25);

  const blockSize = size / qrCode.size;

  for(let x = 0; x < qrCode.size; x++){
    const xPos = x * blockSize;
    for(let y = 0; y < qrCode.size; y++){
      const yPos = y * blockSize;
      if(qrCode.getModule(x, y)){
        renderBlockFunction(xPos, yPos, blockSize);
      }
    }

  }

}


export function renderSwissCross(size: number, renderRectFunction: (x: number, y: number, width: number, height: number, fillColor: string) => void) {

  const scale = size / mm2pt(46);

  const swissCrossWhiteBackgroundSize = mm2pt(7) * scale;
  const swissCrossBlackBackgroundSize = mm2pt(6) * scale;

  const swissCrossThickness = mm2pt(1.17) * scale;
  const swissCrossLength = mm2pt(3.89) * scale;

  renderRectFunction(
    size / 2 - swissCrossWhiteBackgroundSize / 2,
    size / 2 - swissCrossWhiteBackgroundSize / 2,
    swissCrossWhiteBackgroundSize,
    swissCrossWhiteBackgroundSize,
    "white"
  );

  renderRectFunction(
    size / 2 - swissCrossBlackBackgroundSize / 2,
    size / 2 - swissCrossBlackBackgroundSize / 2,
    swissCrossBlackBackgroundSize,
    swissCrossBlackBackgroundSize,
    "black"
  );

  renderRectFunction(
    size / 2 - swissCrossLength / 2,
    size / 2 - swissCrossThickness / 2,
    swissCrossLength,
    swissCrossThickness,
    "white"
  );

  renderRectFunction(
    size / 2 - swissCrossThickness / 2,
    size / 2 - swissCrossLength / 2,
    swissCrossThickness,
    swissCrossLength,
    "white"
  );

}
