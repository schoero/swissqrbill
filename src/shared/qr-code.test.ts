import { describe, expect, test } from "vitest";

import { } from "swissqrbill:shared:cleaner.js";
import { generateQRData } from "swissqrbill:shared:qr-code.js";
import {
  creditorWithBuildingNumber,
  creditorWithBuildingNumberString,
  creditorWithMaxedOutFieldLengths,
  creditorWithNormalIBAN,
  creditorWithNormalIBANAndReference,
  creditorWithQRIBAN,
  debtorWithBuildingNumber,
  debtorWithBuildingNumberString,
  debtorWithMaxedOutFieldLengths,
  minimalRequired,
  minimalRequiredWithAdditionalInformation,
  minimalRequiredWithAlternativeScheme1,
  minimalRequiredWithAlternativeScheme1and2,
  minimalRequiredWithAlternativeScheme2,
  minimalRequiredWithAmount,
  minimalRequiredWithDebtor,
  minimalRequiredWithEuro,
  minimalRequiredWithMaxedOutMessage,
  minimalRequiredWithMessage
} from "swissqrbill:tests:data/valid-data.js";


describe("generateQRData", () => {

  // Minimal required
  test("minimal required data", () => {
    const qrString = generateQRData(minimalRequired);
    expect(qrString).toMatchSnapshot();
  });

  test("minimal required data + amount", () => {
    const qrString = generateQRData(minimalRequiredWithAmount);
    expect(qrString).toMatchSnapshot();
  });

  test("minimal required data + debtor", () => {
    const qrString = generateQRData(minimalRequiredWithDebtor);
    expect(qrString).toMatchSnapshot();
  });

  // Creditor
  test("creditor with building number", () => {
    const qrString = generateQRData(creditorWithBuildingNumber);
    expect(qrString).toMatchSnapshot();
  });

  test("creditor with building number as string", () => {
    const qrString = generateQRData(creditorWithBuildingNumberString);
    expect(qrString).toMatchSnapshot();
  });

  test("creditor with maxed out field lengths", () => {
    const qrString = generateQRData(creditorWithMaxedOutFieldLengths);
    expect(qrString).toMatchSnapshot();
  });

  test("creditor with QR IBAN", () => {
    const qrString = generateQRData(creditorWithQRIBAN);
    expect(qrString).toMatchSnapshot();
  });

  test("creditor with normal IBAN", () => {
    const qrString = generateQRData(creditorWithNormalIBAN);
    expect(qrString).toMatchSnapshot();
  });

  test("creditor with normal IBAN and reference", () => {
    const qrString = generateQRData(creditorWithNormalIBANAndReference);
    expect(qrString).toMatchSnapshot();
  });

  // Debtor
  test("debtor with building number", () => {
    const qrString = generateQRData(debtorWithBuildingNumber);
    expect(qrString).toMatchSnapshot();
  });

  test("debtor with building number as string", () => {
    const qrString = generateQRData(debtorWithBuildingNumberString);
    expect(qrString).toMatchSnapshot();
  });

  test("debtor with maxed out field lengths", () => {
    const qrString = generateQRData(debtorWithMaxedOutFieldLengths);
    expect(qrString).toMatchSnapshot();
  });

  // Message
  test("message", () => {
    const qrString = generateQRData(minimalRequiredWithMessage);
    expect(qrString).toMatchSnapshot();
  });

  test("message with maxed out field length", () => {
    const qrString = generateQRData(minimalRequiredWithMaxedOutMessage);
    expect(qrString).toMatchSnapshot();
  });

  // Currency
  test("currency", () => {
    const qrString = generateQRData(minimalRequiredWithEuro);
    expect(qrString).toMatchSnapshot();
  });

  // Additional Information
  test("additional Information", () => {
    const qrString = generateQRData(minimalRequiredWithAdditionalInformation);
    expect(qrString).toMatchSnapshot();
  });

  // Alternative schemes
  test("alternative schemes AV1", () => {
    const qrString = generateQRData(minimalRequiredWithAlternativeScheme1);
    expect(qrString).toMatchSnapshot();
  });

  test("alternative schemes AV2", () => {
    const qrString = generateQRData(minimalRequiredWithAlternativeScheme2);
    expect(qrString).toMatchSnapshot();
  });

  test("alternative schemes AV1 & AV2", () => {
    const qrString = generateQRData(minimalRequiredWithAlternativeScheme1and2);
    expect(qrString).toMatchSnapshot();
  });

});
