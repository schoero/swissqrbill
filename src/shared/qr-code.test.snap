// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateQRData > additional Information 1`] = `
"SPC
0200
1
CH5800791123000889012
S
Creditor FirstName LastName
Creditor Address

1234
Creditor City
CH







0.00
CHF







NON


EPD
//S1/10/10201409/11/190512/20/1400.000-53/30/106017086/31/180508/32/7.7/40/2:10;0:30"
`;

exports[`generateQRData > alternative schemes AV1 & AV2 1`] = `
"SPC
0200
1
CH5800791123000889012
S
Creditor FirstName LastName
Creditor Address

1234
Creditor City
CH








CHF







NON


EPD

twint/light/02:5d5caa0c078149c694380b72d273ba7#9837183ed9f8bab7286856d786edf5721d55f82b#
rn/twint/a~UuoWrVwETE-AZMysjqoCtQ~s~kNAfGk8vSou0wsvzHvTiSw/rn"
`;

exports[`generateQRData > alternative schemes AV1 1`] = `
"SPC
0200
1
CH5800791123000889012
S
Creditor FirstName LastName
Creditor Address

1234
Creditor City
CH








CHF







NON


EPD

eBill/B/peter@muster.ch"
`;

exports[`generateQRData > alternative schemes AV2 1`] = `
"SPC
0200
1
CH5800791123000889012
S
Creditor FirstName LastName
Creditor Address

1234
Creditor City
CH








CHF







NON


EPD

eBill/B/peter@muster.ch"
`;

exports[`generateQRData > creditor with QR IBAN 1`] = `
"SPC
0200
1
CH4431999123000889012
S
Creditor FirstName LastName
Creditor Address

1234
Creditor City
CH








CHF







QRR
210000000003139471430009017

EPD
"
`;

exports[`generateQRData > creditor with building number 1`] = `
"SPC
0200
1
CH5800791123000889012
S
Creditor FirstName LastName
Creditor Address
123
1234
Creditor City
CH








CHF







NON


EPD
"
`;

exports[`generateQRData > creditor with building number as string 1`] = `
"SPC
0200
1
CH5800791123000889012
S
Creditor FirstName LastName
Creditor Address
A123
1234
Creditor City
CH








CHF







NON


EPD
"
`;

exports[`generateQRData > creditor with maxed out field lengths 1`] = `
"SPC
0200
1
CH5800791123000889012
S
Creditor FirstName LastNameCreditor FirstName LastNameCreditor FirstNa
Creditor AddressCreditor AddressCreditor AddressCreditor AddressCredit

1234123412341234
Creditor CityCreditor CityCreditor 
CH








CHF







NON


EPD
"
`;

exports[`generateQRData > creditor with normal IBAN 1`] = `
"SPC
0200
1
CH5800791123000889012
S
Creditor FirstName LastName
Creditor Address

1234
Creditor City
CH








CHF







NON


EPD
"
`;

exports[`generateQRData > creditor with normal IBAN and reference 1`] = `
"SPC
0200
1
CH5800791123000889012
S
Creditor FirstName LastName
Creditor Address

1234
Creditor City
CH








CHF







SCOR
RF485000056789012345

EPD
"
`;

exports[`generateQRData > currency 1`] = `
"SPC
0200
1
CH5800791123000889012
S
Creditor FirstName LastName
Creditor Address

1234
Creditor City
CH








EUR







NON


EPD
"
`;

exports[`generateQRData > debtor with building number 1`] = `
"SPC
0200
1
CH5800791123000889012
S
Creditor FirstName LastName
Creditor Address

1234
Creditor City
CH








CHF
S
Debtor FirstName LastName
Debtor Address
123
5678
Debtor City
CH
NON


EPD
"
`;

exports[`generateQRData > debtor with building number as string 1`] = `
"SPC
0200
1
CH5800791123000889012
S
Creditor FirstName LastName
Creditor Address

1234
Creditor City
CH








CHF
S
Debtor FirstName LastName
Debtor Address
A123
5678
Debtor City
CH
NON


EPD
"
`;

exports[`generateQRData > debtor with maxed out field lengths 1`] = `
"SPC
0200
1
CH5800791123000889012
S
Creditor FirstName LastName
Creditor Address

1234
Creditor City
CH








CHF
S
Debtor FirstName LastNameDebtor FirstName LastNameDebtor FirstName Las
Debtor AddressDebtor AddressDebtor AddressDebtor AddressDebtor Address

5678567856785678
Debtor CityDebtor CityDebtor CityDe
CH
NON


EPD
"
`;

exports[`generateQRData > message 1`] = `
"SPC
0200
1
CH5800791123000889012
S
Creditor FirstName LastName
Creditor Address

1234
Creditor City
CH








CHF







NON

DO NOT USE FOR PAYMENT
EPD
"
`;

exports[`generateQRData > message with maxed out field length 1`] = `
"SPC
0200
1
CH5800791123000889012
S
Creditor FirstName LastName
Creditor Address

1234
Creditor City
CH








CHF







NON

MessageMessageMessageMessageMessageMessageMessageMessageMessageMessageMessageMessageMessageMessageMessageMessageMessageMessageMessageMessage
EPD
"
`;

exports[`generateQRData > minimal required data + amount 1`] = `
"SPC
0200
1
CH5800791123000889012
S
Creditor FirstName LastName
Creditor Address

1234
Creditor City
CH







123.45
CHF







NON


EPD
"
`;

exports[`generateQRData > minimal required data + debtor 1`] = `
"SPC
0200
1
CH5800791123000889012
S
Creditor FirstName LastName
Creditor Address

1234
Creditor City
CH








CHF
S
Debtor FirstName LastName
Debtor Address

5678
Debtor City
CH
NON


EPD
"
`;

exports[`generateQRData > minimal required data 1`] = `
"SPC
0200
1
CH5800791123000889012
S
Creditor FirstName LastName
Creditor Address

1234
Creditor City
CH








CHF







NON


EPD
"
`;
